### PR TITLE
[docs] overview of cryptographic primitives in Move

### PR DESCRIPTION
**Pull request by alinush**
### Description

**Please DO NOT merge this after approving as I might want to do some more edits!**

1. A long overdue overview of cryptography in Move.
2. Some fixes to the spellchecking script:
   - testing for `aspell` being installed
   - adding support for UTF-8

### CODE REVIEW

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (mstraka100) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:161]: Nit, but having "only sent" here makes me think the received amount isn't hidden. I think the next line makes it clear the system isn't anonymous.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:96]: Suggestion please?

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (mstraka100) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:161]: Sure

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (mstraka100) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:96]: "**any** curve that is or will be supported in the future" in the next line (97)

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (mstraka100) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:182]: Would prefer "We don't currently have an implementation but the reader is encouraged to think of one." I think the implication is too unclear otherwise (that we don't have an example to link to)

**Review by zjma — APPROVED** [review]

File: aptos-move/move-examples/drand/sources/drand.move
Comment (davidiw) [inline aptos-move/move-examples/drand/sources/drand.move:25]: we should probably just move this disclaimer to the README for the entirety of move-examples :grimacing:

File: developer-docs-site/docs/move/move-on-aptos.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos.md:39]: let's put this above resource accounts

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:16]: This feels superfluous. Though I am proud of the work made to get us to where we are today.

Comment (davidiw) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:16]: Maybe just say...

The Aptos adapter for Move has introduced many useful cryptographic primitives. The key guiding principles behind this are

(I don't think we need bold / emphasis)

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:8]: maybe start with a high level of the cryptographic functions we introduce into move followed by the principles

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:40]: typesafety seems to be a core priniciple.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:164]:
```suggestion
**WARNING:** This module is educational. It is **NOT** production-ready. Using it could lead to loss of funds.
```

Comment (davidiw) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:164]: the `likely` is a rather damning comment and makes it read like you may have done something nefarious. As this is our repo and we do at least review code with effort, I'd like to tame it down.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:179]: I think we should think about this statement, it has been reviewed, what is an audit but a review? I think this should be a bit more explicit in that this has not been audited by a third-party auditing organization or something along those lines.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:32]: should we make a table here and discuss tradeoff between collision resistance, gas price, storage, etc?

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:45]: maybe we could make a table here on the pros and cons of these different pieces?

I'd love to push users away from using a key algorithm out of convenience and more from whatever education we can provide them.

**Review by davidiw — COMMENTED** [review]
really nice, really nice

**Review by davidiw — CHANGES_REQUESTED** [review]
requesting changes so that I know I have already looked at it ... feel free to dismiss or request review once you've had a chance to evaluate my comments.

File: aptos-move/move-examples/drand/sources/drand.move
Comment (alinush) [reply aptos-move/move-examples/drand/sources/drand.move]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:25]: I will put them in a table, since it looks better anyway!

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:38]: Hm... that would be a rather nuanced analysis. There's a lot of dimensions to analyze for "pros" and "cons."

I can try to think of some relevant things.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:39]: Fixed: added it to key guiding principles.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md]: Fixed.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (saharct) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:7]:
```suggestion

Cryptography plays an integral role in ensuring the security, integrity, confidentiality, and immutability of data in blockchain systems. The Aptos adapter for Move provides developers with an array of cryptographic primitives to cater to this need. This document delves into the cryptographic functionalities offered by Move on Aptos and elucidates the principles that drive their design.

```

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (saharct) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:14]:
```suggestion
## Cryptographic Primitives

Move, through the Aptos adapter, encompasses several fundamental cryptographic tools:

1. Cryptographic Hash Functions – These are algorithms designed to produce a fixed-size output (hash) from variable-size input data. Supported functions include SHA2-256, SHA3-256, Keccak256, and Blake2b-256.
2. Digital Signature Verification – This process ensures data integrity, authentication, and non-repudiation. Supported signature schemes include Ed25519, ECDSA, and BLS.
3. Elliptic Curve Arithmetic – Elliptic curves are used as the foundation for various cryptographic operations, such as generating keys, performing digital signatures, and verifying signatures. Supported curves include Ristretto255 and BLS12-381.
4. Zero-Knowledge Proofs (ZKP) and Zero-Knowledge Range Proofs (ZKRP) – These cryptographic techniques allow a party to prove a statement's truth without revealing the statement itself. Supported methods include Groth16 ZKP verification and Bulletproofs ZKRP verification.
```

Comment (saharct) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:14]: Created a heading for this section to separate it from the intro.
Added some context to each bullet and combined ZKP and ZKRP into one bullet.

@alinush please modify if you disagree with my descriptions.

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (saharct) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:7]: Let's add an intro here

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:3]: let's avoid slugs

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:14]: nice

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:7]: great idea

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (saharct) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:23]:
```suggestion
Three fundamental principles guide the design and integration of the Aptos cryptographic extensions into Move:

1. **Economic Gas Usage** – Striving to minimize gas costs for Move developers by implementing key primitives as [Move native functions](./move/book/functions#native-functions). For example, see the module for [BLS signatures over BLS12-381 elliptic curves](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/cryptography/bls12381.move).
2. **Type-Safe APIs** – Ensuring that APIs are resistant to common mistakes, type-safety enhances code reliability and promotes an efficient development process. For an example, see the [Ed25519 signature module](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/cryptography/ed25519.move))
3. **Empowerment of Developers** – In instances where native functions are unavailable, we empower developers to build their own cryptographic primitives on top of abstract cryptographic building blocks such as `_finite fields_` and `_Abelian groups_`. Refer to the [`aptos_std::crypto_algebra`](https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-stdlib/sources/cryptography/crypto_algebra.move) module for more insights.

Continue reading to delve a bit deeper and uncover some of the intricacies behind these extensions, as well as the range of applications they empower. For the most comprehensive understanding of this subject, refer to the [cryptography Move modules code](https://github.com/aptos-labs/aptos-core/tree/main/aptos-move/framework/aptos-stdlib/sources/cryptography).
```

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:20]: let's order these a bit differently... maybe type safe APIs is actually the biggest one?

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:35]: could we add in some rough gas usage per ... 1KB or something?
and maybe even potential applications?

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (davidiw) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:52]: please define all acronyms (DLA, ROM)

**Review by davidiw — APPROVED** [review]
unblocking

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (saharct) [inline developer-docs-site/docs/move/move-on-aptos/cryptography.md:72]:
```suggestion
The right choice of a signature scheme in your dapp could depend on many factors:
1. **Backwards-compatibility** 
   - If your dapp's user base predominantly uses a particular signing mechanism, it would be prudent to support that mechanism for ease of transition and adoption.
     - Example: If users mainly sign using Ed25519, it becomes a logical choice.
2. **Ease-of-implementation**
   - While theoretically sound, complex protocols may be challenging to implement in practice.
     - Example: Even though $t$-out-of-$n$ threshold protocols for Ed25519 exist, their intricacy on the signer's side might push developers toward MultiEd25519 due to its more straightforward signing implementation.
3. **Efficiency** 
   - Depending on the dapp's requirements, you might prioritize one aspect of efficiency over another.
     - Signature Size vs. Public Key (PK) Size: Some applications might prioritize a smaller signature footprint, while others might emphasize a compact PK size.
     - Signing Time vs. Verification Time: For certain dapps, the signing speed might be more crucial, while for others, rapid signature verification could be the priority.
4. **Security analysis**
   - It is essential to consider the underlying assumptions and potential vulnerabilities of a signature scheme.
     - Example: ECDSA's security is proven under strong assumptions such as the Generic Group Model (GGM).
     - Malleability Concerns: Some signature schemes are susceptible to malleability, where a valid signature, `$\sigma$`, can be mauled into a different yet still valid signature, `$\sigma'$`, for the same message, `$m$`.
5. **Versatility**
   - The adaptability and flexibility of signature schemes are important to consider so you may properly accommodate the cryptographic needs of your dapp.
     - Example: The $t$-out-of-$n$ threshold BLS is very simple to implement.
```

Comment (saharct) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:72]: I added a short description for each factor and reformatted the examples for improved readability.

**Review by saharct — DISMISSED** [review]
Great stuff, @alinush! Please review these suggestions and the ones I made earlier before you merge. Thanks!

File: developer-docs-site/docs/move/move-on-aptos/cryptography.md
Comment (alinush) [reply developer-docs-site/docs/move/move-on-aptos/cryptography.md:37]: We will do this later, once @mstraka100 figures out an easy way to benchmark gas costs.

**Pull request merged**
